File: packages/loom/parser/src/source.rs
//! Source positions and spans.
//!
//! Loom files are UTF-8 text. Positions carry zero-based line + UTF-8
//! byte offsets — column is the 0-based UTF-8 *byte* column within
//! the line, not a grapheme column. The LSP layer converts to UTF-16
//! at the protocol boundary; everything inside the parser stays in
//! bytes for cheap slicing.

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize, Deserialize)]
pub struct Position {
    pub line: u32,
    pub column: u32,
    pub byte: u32,
}

impl Position {
    pub const fn new(line: u32, column: u32, byte: u32) -> Self {
        Self { line, column, byte }
    }
}

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize, Deserialize)]
pub struct Span {
    pub start: Position,
    pub end: Position,
}

impl Span {
    pub const fn new(start: Position, end: Position) -> Self {
        Self { start, end }
    }

    /// Span covering a single line from `start` to its end column.
    pub const fn line(start: Position, end_byte: u32, end_column: u32) -> Self {
        Self {
            start,
            end: Position {
                line: start.line,
                column: end_column,
                byte: end_byte,
            },
        }
    }
}

File: packages/loom/core/src/parser/source.ts
//! Source positions and spans.
//!
//! Positions carry a zero-based line plus a JS-string offset (UTF-16
//! code units). `column` is the offset within the line; `offset` is the
//! offset from the start of the file. For ASCII source these match the
//! Rust parser's UTF-8 byte offsets exactly; the field is named
//! `offset` (not `byte`) because the unit is a JS string index.

export interface Position {
  line: number;
  column: number;
  offset: number;
}

export function pos(line: number, column: number, offset: number): Position {
  return { line, column, offset };
}

export const ZERO_POSITION: Position = { line: 0, column: 0, offset: 0 };

export interface Span {
  start: Position;
  end: Position;
}

export function span(start: Position, end: Position): Span {
  return { start, end };
}

export const ZERO_SPAN: Span = { start: ZERO_POSITION, end: ZERO_POSITION };

/** Span covering a single line from `start` to its end column/offset. */
export function lineSpanOf(start: Position, endOffset: number, endColumn: number): Span {
  return {
    start,
    end: { line: start.line, column: endColumn, offset: endOffset },
  };
}
